src/git/repo.rs

```diff
@@ -965,6 +1022,144 @@ Either create it, or update the main branch setting by running:
         Ok(make_non_zero_oid(oid))
     }
+
+    /// Amends the provided parent commit in memory and returns the resulting tree.
+    ///
+    /// Only amends the files provided in the options, and only supports amending from
+    /// either the working tree or the index, but not both.
+    ///
+    /// See `Repo::cherry_pick_fast` for motivation for performing the operation
+    /// in-memory.
+    pub fn amend_fast(
+        &self,
+        parent_commit: &Commit,
+        opts: &AmendFastOptions,
+    ) -> eyre::Result<Tree> {
+        let parent_commit_pathbufs = self
+            .get_paths_touched_by_commit(parent_commit)?
+            .ok_or_else(|| {
+                eyre::eyre!(
+                    "Could not get paths touched by commit: {:?}",
+                    &parent_commit
+                )
+            })?
+            .into_iter()
+            .collect_vec();
+        let changed_paths = {
+            let mut result: HashSet<&Path> =
+                parent_commit_pathbufs.iter().map(PathBuf::borrow).collect();
+            let paths = match opts {
+                AmendFastOptions::FromIndex { paths } => paths,
+                AmendFastOptions::FromWorkingTree { paths } => paths,
+            };
+            result.extend(paths.iter().map(|path| path.as_path()));
+            result.iter().cloned().collect_vec()
+        };
+
+        let dehydrated_parent =
+            self.dehydrate_commit(parent_commit, changed_paths.as_slice(), true)?;
+        let dehydrated_parent_index = Index::from_tree(&dehydrated_parent.get_tree()?)?;
+
+        let repo_path = self
+            .get_working_copy_path()
+            .expect("repo working copy path");
+        let amended_oids: HashMap<PathBuf, Option<NonZeroOid>> = match opts {
+            AmendFastOptions::FromWorkingTree { paths } => paths
+                .iter()
+                .map(|path| {
+                    let file_path = &repo_path.join(path);
+                    if file_path.exists() {
+                        // Create a new blob OID based on the current on-disk
+                        // contents of the file in the working copy.
+                        self.inner
+                            .blob_path(file_path)
+                            .map(|oid| (path.clone(), Some(make_non_zero_oid(oid))))
+                    } else {
+                        Ok((path.clone(), None))
+                    }
+                })
+                .collect::<Result<HashMap<_, _>, _>>()?,
+            AmendFastOptions::FromIndex { paths } => {
+                let index = self.get_index()?;
+                paths
+                    .iter()
+                    .filter_map(|path| match index.get_entry(path) {
+                        Some(IndexEntry {
+                            oid: MaybeZeroOid::Zero,
+                            ..
+                        }) => {
+                            warn!("index entry for {:?} was zero", path);
```

Comment:
Use `warn!(?path, "index entry was zero")` to get the structured logging equivalent, which can be analyzed programmatically.